### ISSUE TITLE
CFArray::from_copyable can corrupt memory

### ISSUE THREAD

**Issue by mehcode** [servo/core-foundation-rs#291]
https://github.com/servo/core-foundation-rs/blob/36d7c972feeeb4a93879e76ac0b9d410cb6130b8/core-foundation/src/array.rs#L66-L74

---

Elements in CFArray are pointers. It should be ok to allow storing raw pointers, `usize`, etc. in the CFArray, but not any `Copy`-able.

**Comment by jdm** [comment]
Could you explain how memory can be corrupted? https://developer.apple.com/documentation/corefoundation/1388741-cfarraycreate?language=objc says "This C array is not changed or freed by this function." and "The number of values to copy from the values C array into the new array."

**Comment by jdm** [comment]
Oh, I just noticed "A C array of the pointer-sized values to be in the new array.". That is one way in which the resulting array's contents would not necessarily match the original slice's contents.

**Comment by mehcode** [comment]
Sorry, should have clarified. I didn't mean that specific function is corrupting anything but the array itself.

But combined with the accessor functions this could be dangerous.

**Comment by application-developer-DA** [comment]
Indeed. I encountered a weird bug yesterday and spent several hours debugging my code, but then I realized that the issue is in `CFArray::from_copyable()`.

A simple example which reproduces this issue:
```
use core_foundation::array::CFArray;
use core_graphics::display::CGWindowID;

fn main() {
    let windows: Vec<CGWindowID> = vec![1738, 1416];
    println!("Original: {:?}", windows);

    let array = CFArray::from_copyable(&windows);
    let untyped = array.to_untyped();
    for value in untyped.get_all_values() {
        println!("After Getting Values: {}", value as u32);
    }
}
```

The output is:
```
Original: [1738, 1416]
After Getting Values: 1738
After Getting Values: 3775661375
```

(Sometimes the second value is `0`, sometimes some random number.)

I encountered this issue when I was using `CGDisplay::screenshot_from_windows()` which accepts a `CFArray` which I constructed from a vector of `CGWindowID` using `CFArray::from_copyable()`.

Perhaps we have to make `CFArray::from_copyable()` an unsafe function and document its usage. Another way would be to use a concrete type for `CGDisplay::screenshot_from_windows()` and similar functions, so that the user does not have to guess, how the underlying low level function interprets the pointers (whehter it just performs a cast or tries to access the value to which the pointer points).

**Comment by jrmuizel** [comment]
We should just get rid of `CFArray::from_copyable()`. The correct solution here is to have a trait like `PointerSized` implement that for a number of things and add something like `CFArray::from_pointer_sized()`

**Comment by roblabla** [comment]
`from_copyable` isn't even used in core-graphics, which according to the commit (https://github.com/servo/core-foundation-rs/commit/90ddbb60b241bd6043f60685115591c8f5307970) is what it was used for. Besides, it's not even valid for the intended use-case: CGWindowID is 32-bit, and not pointer sized!

The function isn't used anywhere from a cursory github-wide search. I think it should get removed entirely and not replaced without a valid use-case.
